common/orcahobbit.1
.TH ORCAHOBBIT 1 "Version 4.2.2: 15 Dec 2008" "Xymon"
.SH NAME
orcahobbit \- Xymon client utility to grab data from ORCA
.SH SYNOPSIS
.B "orcahobbit --orca=PREFIX [options]"

.SH NOTICE
This utility is included in the client distribution for Xymon 4.2.
However, the backend module to parse the data it sends it \fBNOT\fR
included in Xymon 4.2. It is possible to use the generic Xymon
NCV data handler in
.I hobbitd_rrd(8)
to process ORCA data, if you have an urgent need to do so.

.SH DESCRIPTION
\fBorcahobbit\fR is an add-on tool for the Xymon client. It is 
used to grab data collected by the ORCA data collection tool 
(orcallator.se), and send it to the Xymon server in NCV format.

orcahobbit should run from the client
.I hobbitlaunch(8)
utility, i.e. there must be an entry in the 
.I clientlaunch.cfg(5)
file for orcahobbit.

.SH OPTIONS
.IP "--orca=PREFIX"
The filename prefix for the ORCA data log. Typically this is the
directory for the ORCA logs, followed by "orcallator". The actual
filename for the ORCA logs include a timestamp and sequence number, 
e.g. "orcallator-2006-06-20-000". This option is required.

.IP "--debug"
Enable debugging output.

.SH "SEE ALSO"
xymon(7), clientlaunch.cfg(5)



Changes
Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------
r5940 | storner | 2008-11-26 13:17:12 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitfetch.c

hobbitfetch insisted that hosts must have a valid IP-adress in bb-hosts. This causes problems for hosts with dynamic IP-adresses. This patch causes hobbitfetch to lookup the IP-address at run-time if the IP is listed as "0.0.0.0".


------------------------------------------------------------------------
r5941 | storner | 2008-11-26 13:20:44 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbitsvc-info.c

Hostname filtering on the info column page is broken, if you have multiple hosts with similar names, resulting in the same columns showing up multiple times in the disable section. A similar problem affects host filtering in the configuration report tool.


------------------------------------------------------------------------
r5942 | storner | 2008-11-26 13:21:27 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/hobbitclient-sunos.sh

The Solaris client would not collect the "top" process-listing data which is normally included in the data on the "cpu" column.


------------------------------------------------------------------------
r5943 | storner | 2008-11-26 13:25:28 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/build/bb-commands.sh
   M /branches/4.2.0/client/hobbitclient-sunos.sh

The Solaris lofs (loopback filesystem) should not be included in the df output appearing on the Disk status.
------------------------------------------------------------------------
r5944 | storner | 2008-11-26 13:29:56 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/lib/cgi.c
   M /branches/4.2.0/lib/cgi.h
   M /branches/4.2.0/web/bb-ack.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-hostgraphs.c
   M /branches/4.2.0/web/hobbit-statusreport.c

Cookie handling does not always work. This results in hosts not showing up on the new acknowledgment page, or hosts missing from the Metrics report.
------------------------------------------------------------------------
r5945 | storner | 2008-11-26 13:30:31 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/bbdisplay/bbgen.h

Hobbit 4.2.0 fails to build on some OS X releases due to a naming conflict between a system-defined datatype, and a different datatype defined in Hobbit.
------------------------------------------------------------------------
r5946 | storner | 2008-11-26 13:31:41 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/hobbitd/webfiles/zoom.js
   M /branches/4.2.0/lib/hobbitrrd.c
   M /branches/4.2.0/lib/hobbitrrd.h
   M /branches/4.2.0/lib/htmllog.c
   M /branches/4.2.0/web/hobbitgraph.c
   M /branches/4.2.0/web/hobbitsvc-trends.c

When viewing detailed graphs linked from the detailed status page, the background color of the webpages would always show green, regardless of the actual color of the status it was linked from. This patch keeps the background color at what the status page shows.
------------------------------------------------------------------------
r5947 | storner | 2008-11-26 13:32:55 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/hobbitd/webfiles/confreport_front
   M /branches/4.2.0/hobbitd/wwwfiles/menu/menu_items.js.DIST
   M /branches/4.2.0/web/Makefile
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbit-confreport.cgi.1

The Configuration Report did not include information about what systems appears on the "Critical Systems" view.

client/orcahobbit.c
/*----------------------------------------------------------------------------*/
/* Hobbit ORCA data collector.                                                */
/* This tool grabs the last reading from an ORCA logfile and formats it in    */
/* NAME:VALUE format for the client message.                                  */
/*                                                                            */
/* Copyright (C) 2006-2008 Henrik Storner <henrik@hswn.dk>                    */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id$";

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/time.h>
#include <time.h>
#include <limits.h>

#include "libbbgen.h"

int main(int argc, char *argv[])
{
	time_t now;
	char *prefix = NULL, *machinename = NULL;
	char datestr[12], fn[PATH_MAX];
	int i;
	FILE *fd;
	char headerline[32768];
	char vals[32768];
	int gotvals = 0;
	char *hp, *hdr, *vp, *val;
	char msgline[4096];
	strbuffer_t *msg;

	machinename = xgetenv("MACHINE");

	for (i=1; (i < argc); i++) {
		if (strncmp(argv[i], "--orca=", 7) == 0) {
			prefix = argv[i]+7;
		}
		else if (strncmp(argv[i], "--machine=", 10) == 0) {
			machinename = argv[i]+10;
		}
		else if (strcmp(argv[i], "--debug") == 0) {
			debug = dontsendmessages = 1;
		}
	}

	if (!prefix || !machinename) return 0;

	/* 
	 * ORCA logfiles are names PREFIX-%Y-%m-%d-XXX where XXX is a 
	 * number starting at 0 and increasing whenever the columns
	 * change.
	 * We will look for the first 20 index numbers only.
	 */
	now = getcurrenttime(NULL);
	strftime(datestr, sizeof(datestr), "%Y-%m-%d", localtime(&now));
	i = 0; fd = NULL;
	while ((i < 20) && !fd) {
		snprintf(fn, sizeof(fn), "%s-%s-%03d", prefix, datestr, i);
		fd = fopen(fn, "r");
	}

	if (!fd) return 1;

	/* Grab the header line, and the last logfile entry. */
	if (fgets(headerline, sizeof(headerline), fd)) {
		while (fgets(vals, sizeof(vals), fd)) gotvals = 1;
	}
	fclose(fd);

	msg = newstrbuffer(0);
	sprintf(msgline, "data %s.orca\n", machinename);
	addtobuffer(msg, msgline);

	/* Match headers and values. */
	hdr = strtok_r(headerline, " \t\n", &hp);
	val = strtok_r(vals, " \t\n", &vp);
	while (hdr && val) {
		sprintf(msgline, "%s:%s\n", hdr, val);
		addtobuffer(msg, msgline);
		hdr = strtok_r(NULL, " \t\n", &hp);
		val = strtok_r(NULL, " \t\n", &vp);
	}

	sendmessage(STRBUF(msg), NULL, BBTALK_TIMEOUT, NULL);

	return 0;
}

